File: solutions/lua/change/2/change.lua
-- my first, recursive version doesn't find the solution with the minimum number of coins 
-- the following is essentially Gelennj's solution
-- (I wanted to understand the algorithm better)

-- the following function finds the solutions for all amounts up to amount the given values (i.e. coin denominations)
-- only S, which stores the first coin idx for every amount is used later

local function change (amount, values)
  local C = {[0] = 0} 
  local S = {}
  for p = 1, amount do
      local min = math.maxinteger
      local coin
      for i = 1, #values do
        if values[i] <= p then
          if C[p - values[i]]<min then
                 min = 1 + C[p - values[i]]
                 coin = i
          end
        end              
      end
      C[p] = min
      S[p] = coin
  end
  return C,S
end

-- now, starting with n, we retrive the solution from S

local function make_solution (S, values, n)
    result = {}
    while n>0 do
      local coin = values[S[n]] 
      assert(coin, "can't make target with given coins")
      table.insert(result, 1, coin)
      n = n - coin
    end
    return result
end

return function (amount, values)
  assert(amount>=0, "target can't be negative") 
  if amount==0 then 
      return {}
  end
  _, first_cs = change(amount, values)
  return make_solution(first_cs, values, amount)  
end   

File: solutions/julia/change/1/change.jl
# solutions finds the correct starting coin for every amount 
# between 1 and target 

function solutions(coins, target)
    c_nums = Dict(0=>0) # min no. of coins for each amount
    sol_indices = []    # starting coin for each amount
    for p in 1:target 
        coin = 0
        min = typemax(Int64)
        for i in 1:length(coins)
            if coins[i]<=p 
                if c_nums[p-coins[i]]<min
                    min = 1 + c_nums[p-coins[i]]
                    coin = i
                end
            end
        end
        c_nums[p] =  min
        push!(sol_indices, coin)
    end
    sol_indices
end

# make_solution retives the solution for target n from 
# sols, the vector of first coins to use

function make_solution(sols, coins, n)
    res = []
    while n>0 
        sols[n] !=0 || throw(DomainError(n,"can't make target with given coins"))
        coin = coins[sols[n]]
        push!(res,coin)
        n -= coin
    end
    sort(res)
end

function change(coins, target)
    target>=0 || throw(DomainError(target, "target can't be negative"))
    sols = solutions(coins, target)
    make_solution(sols, coins, target)
end
